src/averageAges.js

```diff
@@ -56,6 +89,42 @@ function calculateWomenAverageAge(people, withChildren) {
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
   // write code here
+  const mothers = people.filter(person => {
+    return people.some(child => child.mother === person.name);
+  });
+
+  const childrens = mothers.map(mother => {
+    const children = people.filter(child => child.mother === mother.name);
+
+    return children;
+  }).reduce((total, current) => total.concat(current), []);
+
+  const ageDifferences = childrens.map(child => {
+    const childMother = mothers.find(mother => mother.name === child.mother);
+
+    return child.born - childMother.born;
+  });
+
+  const boys = childrens.filter(child => child.sex === 'm');
+
+  const ageDiffwithBoys = boys.map(boy => {
+    const boyMother = mothers.find(mother => mother.name === boy.mother);
+
+    return boy.born - boyMother.born;
+  });
+
+  const sumOfAge = ageDifferences.reduce((total, current) => {
+    return total + current;
+  }, 0);
+
+  const sumOfAgeWithBoys = ageDiffwithBoys.reduce((total, current) => {
+    return total + current;
+  }, 0);
+
+  const avgAgeDifference = sumOfAge / ageDifferences.length;
+  const avgAgeDiffWithBoys = sumOfAgeWithBoys / boys.length;
+
+  return onlyWithSon ? avgAgeDiffWithBoys : avgAgeDifference;
```

Comment:
Similar to function 1, first determine the group of people and then count the age difference just once